docs/DesignDocs/Attestation_API_Proposal.md

```diff
@@ -0,0 +1,534 @@
+Proposal of OE SDK Attestation Public and Plugin API for V0.10 Release
+====
+
+# Background
+
+This document describes the proposed OE SDK attestation public and plugin API
+for V0.10 release, as an evolution of the experimental attestation API
+implemented in the
+[OE SDK V0.9 release](https://github.com/openenclave/openenclave/tree/v0.9.x).
+
+This proposal is based on these efforts:
+- Design document [Attestation V3 Update]:
+[Custom Attestation Data Formats for Open Enclave V3 Updates](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/CustomAttestation_V3.md).
+- Discussion in [Issue #2729]: [Outstanding issues on attestation plugin design](https://github.com/openenclave/openenclave/issues/2729).
+- Design document [quote-ex Integration]:
+[Attestation: OE SDK Integration with Intel® SGX SDK quote-ex Library for Generation of Evidence in New Formats](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/SGX_QuoteEx_Integration.md).
+- Design document [Design Notes]:
+[Notes on OE SDK Attestation API and Plugin Library Design](https://github.com/shnwc/openenclave/blob/master/docs/DesignDocs/NotesOnAttestationAPI.md).
+  - Note: this document is in the process of being up-streamed as OE SDK [PR #2801](https://github.com/openenclave/openenclave/pull/2801).
+
+# OE SDK Attestation Public API Proposal
+
+## Existing OE SDK V0.9 Public API
+
+The OE SDK V0.9 experimental public API for attestation is composed of
+functions for a number of use cases to be implemented by applications.
+The use cases fall into two categories: evidence generation and verification.
+
+Use cases for evidence generation:
+
+- Retrieve a single SGX attester plugin.
+  - Function `oe_attester_t* oe_sgx_plugin_attester(void)`,
+  declared in header file [openenclave/attestation/sgx/attester.h](https://github.com/openenclave/openenclave/blob/v0.9.x/include/openenclave/attestation/sgx/attester.h).
+  - It returns a single hardcoded SGX attester plugin for generation of
+  evidences in SGX ECDSA-p256 (simply called ECDSA) format.
+- Register an attester plugin.
+  - Function `oe_result_t oe_register_attester(oe_attester_t* plugin, const void* config_data, size_t config_data_size)`, declared in header file [openenclave/attestation/plugin.h](https://github.com/openenclave/openenclave/blob/v0.9.x/include/openenclave/attestation/plugin.h).
+  - The attester plugin returned by `oe_sgx_plugin_attester()` can be
+  registered with this function.
+  - Note: requirement for the application to supply a plugin-specific `config_data`
+  parameter is against the goal of a TEE-agnostic OE SDK API. More discussion on
+  TEE agnostic design can be found in the
+  [[Design Notes]](https://github.com/shnwc/openenclave/blob/master/docs/DesignDocs/NotesOnAttestationAPI.md#attester--verifier-security-model-and-tee-agnostic-design)
+  document.
+- Get an evidence in a globally unique format, optionally along with a set of
```

Comment:
```suggestion
- Get evidence in a globally unique format, optionally along with a set of
```

Reply:
Thanks for the catch! While "evidence" is uncountable, what I see here is something countable: every call to `oe_get_evidence()` returns a different piece of evidence. Maybe should replace "an evidence" with "a piece of evidence"?

Reply:
Per the rats architecture, a piece of evidence would be a specific claim.   A set of claims is what gets signed (at least typically), and the signed set of claims is called evidence.

Reply:
Got it. So we stick with the uncountable "evidence".